### PR TITLE
Snow Leopards - Alaere N. 

### PR DESCRIPTION
**Pull request by Alaere00**

### CODE REVIEW

File: test/adagrams.test.js [hunk 1]
```diff
@@ -87,7 +87,7 @@ describe("Adagrams", () => {
       expect(isValid).toBe(false);
     });
 
-    it("returns false when word contains repeated letters more than in the drawn letters", () => {
+    it("returns true when word contains repeated letters more than in the drawn letters", () => {
```
Comment (anselrognlie) [inline test/adagrams.test.js:90]: I think the name of this test was accurate for the test. We do want this to return false if the word uses more copies of a letter than is available in the hand.

File: test/adagrams.test.js [hunk 2]
```diff
@@ -120,8 +120,12 @@ describe("Adagrams", () => {
     });
 
     it("returns a score of 0 if given an empty input", () => {
-      throw "Complete test";
+      expectScores({
+        '': 0,
+        PANCAKE: 23,
+        CAT: 5
```
Comment (anselrognlie) [inline test/adagrams.test.js:126]: The empty string key is what we'd be looking for here. The other cases are great if you want to use them for testing, but they would be better added to the earlier test `returns an accurate numerical score according to the score chart` rather than added here, since this test is named `returns a score of 0 if given an empty input`.

File: test/adagrams.test.js [hunk 3]
```diff
@@ -145,7 +149,7 @@ describe("Adagrams", () => {
       const words = ["XXX", "XXXX", "X", "XX"];
       const correct = { word: "XXXX", score: scoreWord("XXXX") };
 
-      throw "Complete test by adding an assertion";
+      expect(highestScoreFrom(words)).toEqual(correct);
```
Comment (anselrognlie) [inline test/adagrams.test.js:152]: 👍

File: src/adagrams.js [hunk 4]
```diff
@@ -1,15 +1,112 @@
+const letterBank = {
```
Comment (anselrognlie) [inline src/adagrams.js:1]: I appreciate the desire to combine these into a single structure, but by using a list as the value, it's difficult to know what the two numbers are doing. Later, when you acces position 0 of the list, what does that mean? It;s hard to tell from the code. We could make it more clear by using objects with keys (like maybe pool and points) as the value, though then we're doing a lot of typing. So then we might make a helper function to build that object for us.

Or, we could store the values in two different collections where the collection names describe what the values would be.

File: src/adagrams.js [hunk 5]
```diff
@@ -1,15 +1,112 @@
...
+  let deck = [];
+
+  for (const [key, cnt] of Object.entries(letterBank)){
+    for (let i=0; i<cnt[0]; i++){
+      deck.push(key);
+    }
+  }
+  let dCnt = deck.length - 1
+  while (dCnt) {
+    let i = Math.floor(Math.random() * dCnt--);
```
Comment (anselrognlie) [inline src/adagrams.js:49]: This is where ++/-- come into their own. We can get the value of the variable, and update it. Some teams may not like this, as it's easy to overlook. But it can be especially nice when the ++/-- is part of the while condition itself, since then it's more clear that the check and the update are happening together.

File: src/adagrams.js [hunk 6]
```diff
@@ -1,15 +1,112 @@
...
-  // Implement this method for wave 1
+  let deck = [];
+
+  for (const [key, cnt] of Object.entries(letterBank)){
+    for (let i=0; i<cnt[0]; i++){
+      deck.push(key);
+    }
+  }
+  let dCnt = deck.length - 1
+  while (dCnt) {
```
Comment (anselrognlie) [inline src/adagrams.js:48]: Nice shuffle implementation. Consider moving it to a helper function to help readability.

File: src/adagrams.js [hunk 7]
```diff
@@ -1,15 +1,112 @@
...
+    let i = Math.floor(Math.random() * dCnt--);
+
+    [deck[dCnt], deck[i]] = [deck[i], deck[dCnt]];
+
+  }
+  return deck.splice(0, 10);
```
Comment (anselrognlie) [inline src/adagrams.js:54]: `splice` is more commonly used when the intent is to modify the array it's being called on (like `remove` does in python). If all we need is the subset of the array from 0 to 10, and don't need to modify the underlying array (which has to shift the array contents), we can use `slice` instead.

File: src/adagrams.js [hunk 8]
```diff
@@ -1,15 +1,112 @@
...
+  U: [4, 1],
+  V: [2, 4],
+  W: [2, 4],
+  X: [1, 8],
+  Y: [2, 4],
+  Z: [1, 10],
+};
+const counter = array => {
+  let count = {};
+  array.forEach(val => count[val] = (count[val] || 0) + 1);
```
Comment (anselrognlie) [inline src/adagrams.js:31]: Nice use of `||` to stand in for `get` in python. But notice how when we try to write code more conscisely, which functional styles seem to drive us towards even more, we can wind up with code that looks like a random assortment of symbols, which can be tough to decipher. You helped with that by sequestering it in a helper method.

File: src/adagrams.js [hunk 9]
```diff
@@ -1,15 +1,112 @@
...
+    [deck[dCnt], deck[i]] = [deck[i], deck[dCnt]];
+
+  }
+  return deck.splice(0, 10);
 };
 
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
-};
+  let objLInHand = counter(lettersInHand);
```
Comment (anselrognlie) [inline src/adagrams.js:58]: Nice helper to calculate a frequency map of the letters in the hand.

File: src/adagrams.js [hunk 10]
```diff
@@ -1,15 +1,112 @@
...
 
-export const scoreWord = (word) => {
-  // Implement this method for wave 3
+  for (let l of input){
+    if (l in objLInHand && objLInHand[l] > 0){
+      objLInHand[l] -= 1;
+      continue
```
Comment (anselrognlie) [inline src/adagrams.js:63]: Consider omitting the `continue`. There's no code that it's skipping, so this makes the reader spendtime questioning why it's here.

File: src/adagrams.js [hunk 11]
```diff
@@ -1,15 +1,112 @@
...
 
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
-};
+  let objLInHand = counter(lettersInHand);
 
-export const scoreWord = (word) => {
-  // Implement this method for wave 3
+  for (let l of input){
+    if (l in objLInHand && objLInHand[l] > 0){
```
Comment (anselrognlie) [inline src/adagrams.js:61]: We could omit the `in` check. JS returns `undefined`, which is not greater than 0. So if a key is missing, we would hit the case hat returns `false`, which is what we would want.

File: src/adagrams.js [hunk 12]
```diff
@@ -1,15 +1,112 @@
...
+      return false;
+    }
+  }
+  return true;
 };
+export const scoreWord = (word) => {
+  word = word.toUpperCase();
+  let score = 0;
 
+  for (let i = 0; i<word.length; ++i){
```
Comment (anselrognlie) [inline src/adagrams.js:74]: Prefer a `for/of`
```js
  for (const letter of word){
    // use letter in place of word[i] below
```

File: src/adagrams.js [hunk 13]
```diff
@@ -1,15 +1,112 @@
...
+      score += letterBank[word[i]][1];
```
Comment (anselrognlie) [inline src/adagrams.js:75]: What's stored in index 1 here? The score for the letter. But we can't really tell since we're accessing it by the offset alone.

File: src/adagrams.js [hunk 14]
```diff
@@ -1,15 +1,112 @@
...
+  Z: [1, 10],
+};
+const counter = array => {
+  let count = {};
+  array.forEach(val => count[val] = (count[val] || 0) + 1);
+  return count;
+};
+
+const shortestWord = (array) => {
+  array.sort((a, b) => a.length - b.length);
```
Comment (anselrognlie) [inline src/adagrams.js:36]: Finding the shortest thing in a list is more directly found by iterating over the list and tracking the shortest length and the associated word (O(n)). Sorting is always at least O(n log n).

For a more functional approach, we can use `reduce` to find min/max values if we're careful with our reduce condition
```js
  return array.reduce((shortest, current) => current.length < shortest.length ? current : shortest).length;
```

File: src/adagrams.js [hunk 15]
```diff
@@ -1,15 +1,112 @@
...
+  Y: [2, 4],
+  Z: [1, 10],
+};
+const counter = array => {
+  let count = {};
+  array.forEach(val => count[val] = (count[val] || 0) + 1);
+  return count;
+};
+
+const shortestWord = (array) => {
```
Comment (anselrognlie) [inline src/adagrams.js:35]: Consider naming this `shortestWordLength`. `shortestWord` sounds like it will return the actual word rather than the length.

File: src/adagrams.js [hunk 16]
```diff
@@ -1,15 +1,112 @@
...
+  const minLength = shortestWord(words);
+
+  for (const [word, score] of Object.entries(scoreBoard)){
+    if (score === maxScore){
+      highScores.push(word);
+    } 
+  };
+  if (highScores.length === 1){
+    return {word: highScores[0], score:maxScore};
+  } else if (highScores.length > 1){
```
Comment (anselrognlie) [inline src/adagrams.js:98]: This could possibly be a plain `else`. The only other condition might be if there were no words supplied, but we have no tests for that. Arguably, if the caller can tell that there were no words (a round was played and not player was able to make a word), it really shouldn't be calling this function!

File: src/adagrams.js [hunk 17]
```diff
@@ -1,15 +1,112 @@
...
+    for (const word of highScores){
+      if (word.length === 10){
+        return {word, score:maxScore};
+      } else {
+        const wordLength = highScores.every( w => w.length < 10);
+        if (wordLength === true && word.length === minLength){
+        return {word, score:maxScore};
```
Comment (anselrognlie) [inline src/adagrams.js:105]: Consider doing a second iteration after looping through looking for a 10-letter word, looking the shortest word. `every` is doing a loop within a loop for each inspected word. But if we wait until this first loop finishes, then we'll know that all of the words are less than 10 letters (or we'd have returned), removing the need for the `every` check.

Comment (anselrognlie) [reply src/adagrams.js:105]: There's no guarantee that the minimum length word is in the tied words (since it was calculated from the entire word list). We should wait to find the shortest word until we know which words are in the tied list.

**Review by anselrognlie — COMMENTED** [review]
Looks good! You used a lot of JS functionality, and built the additional functionality that you needed. I see a healthy mixture of code that might feel more at home in Python, but also some approaches that feel very JS. Your tests are all passing, but take a look at my comments on Wave 4. There might be a corner case to reconsider. Nicely done!
